internal/pfcpsim/server.go

```diff
@@ -0,0 +1,316 @@
+
+	infoMsg := "Association teardown completed and connection to remote peer closed"
+	log.Info(infoMsg)
+
+	return &pb.Response{
+		StatusCode: int32(codes.OK),
+		Message:    infoMsg,
+	}, nil
+}
+
+func (P PFCPSimService) CreateSession(ctx context.Context, request *pb.CreateSessionRequest) (*pb.Response, error) {
+	if !isConfigured() {
+		return &pb.Response{}, status.Error(codes.Aborted, "Server is not configured")
+	}
+
+	baseID := int(request.BaseID)
+	count := int(request.Count)
+
+	lastUEAddr, _, err := net.ParseCIDR(request.UeAddressPool)
+	if err != nil {
+		errMsg := fmt.Sprintf(" Could not parse Address Pool: %v", err)
+		log.Error(errMsg)
+		return &pb.Response{}, status.Error(codes.Aborted, errMsg)
+	}
+
+	nodebAddr := request.NodeBAddress
+
+	for i := baseID; i < (count*2 + baseID); i = i + 2 {
+		// using variables to ease comprehension on how rules are linked together
+		uplinkTEID := uint32(i)
+		downlinkTEID := uint32(i + 1)
+
+		ueAddress := iplib.NextIP(lastUEAddr)
+		lastUEAddr = ueAddress
+
+		uplinkFarID := uint32(i)
+		downlinkFarID := uint32(i + 1)
+
+		uplinkPdrID := uint16(i)
+		dowlinkPdrID := uint16(i + 1)
+
+		sessQerID := uint32(i + 3)
+
+		appQerID := uint32(i)
+
+		uplinkAppQerID := appQerID
+		downlinkAppQerID := appQerID + 1
+
+		pdrs := []*ieLib.IE{
+			// UplinkPDR
+			session.NewPDRBuilder().
+				WithID(uplinkPdrID).
+				WithMethod(session.Create).
+				WithTEID(uplinkTEID).
+				WithFARID(uplinkFarID).
+				AddQERID(sessQerID).
+				AddQERID(uplinkAppQerID).
+				WithN3Address(upfN3Address).
+				WithSDFFilter(defaultSDFfilter).
+				MarkAsUplink().
+				BuildPDR(),
+
+			// DownlinkPDR
+			session.NewPDRBuilder().
+				WithID(dowlinkPdrID).
+				WithMethod(session.Create).
+				WithPrecedence(100).
+				WithUEAddress(ueAddress.String()).
+				WithSDFFilter(defaultSDFfilter).
+				AddQERID(sessQerID).
+				AddQERID(downlinkAppQerID).
+				WithFARID(downlinkFarID).
+				MarkAsDownlink().
+				BuildPDR(),
+		}
+
+		fars := []*ieLib.IE{
+			// UplinkFAR
+			session.NewFARBuilder().
+				WithID(uplinkFarID).
+				WithAction(session.ActionForward).
+				WithDstInterface(ieLib.DstInterfaceCore).
+				WithMethod(session.Create).
+				BuildFAR(),
+
+			// DownlinkFAR
+			session.NewFARBuilder().
+				WithID(downlinkFarID).
+				WithAction(session.ActionDrop).
+				WithMethod(session.Create).
+				WithDstInterface(ieLib.DstInterfaceAccess).
+				WithTEID(downlinkTEID).
+				WithDownlinkIP(nodebAddr).
+				BuildFAR(),
+		}
+
+		qers := []*ieLib.IE{
+			// session QER
+			session.NewQERBuilder().
+				WithID(sessQerID).
+				WithMethod(session.Create).
+				WithQFI(0x09).
+				WithUplinkMBR(50000).
+				WithDownlinkMBR(50000).
+				Build(),
+
+			// application QER
+			session.NewQERBuilder().
+				WithID(appQerID).
+				WithMethod(session.Create).
+				WithQFI(0x08).
+				WithUplinkMBR(50000).
+				WithUplinkGBR(50000).
+				WithDownlinkMBR(30000).
+				WithUplinkGBR(30000).
+				Build(),
```

Comment:
I'm not sure about this. But since we are sending a single PDR, we should have only one QER (namely an application QER). @osinstom, right?

Reply:
As we discussed on Slack we should have 3 QERs:
- 1 session QER
- 1 UL app QER
- 1 DL app QER
UL and DL app QER should be equal.